common/nym-lp/src/session.rs

```diff
@@ -0,0 +1,1898 @@
+    ResponderProcessed,
+}
+
+impl std::fmt::Debug for KKTState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::NotStarted => write!(f, "KKTState::NotStarted"),
+            Self::InitiatorWaiting { context } => f
+                .debug_struct("KKTState::InitiatorWaiting")
+                .field("context", context)
+                .finish(),
+            Self::Completed { .. } => write!(f, "KKTState::Completed {{ kem_pk: <opaque> }}"),
+            Self::ResponderProcessed => write!(f, "KKTState::ResponderProcessed"),
+        }
+    }
+}
+
+/// PSQ (Post-Quantum Secure PSK) handshake state.
+///
+/// Tracks the PSQ protocol state machine through the session lifecycle.
+///
+/// # State Transitions
+///
+/// **Initiator path:**
+/// ```text
+/// NotStarted → InitiatorWaiting → Completed
+/// ```
+///
+/// **Responder path:**
+/// ```text
+/// NotStarted → ResponderWaiting → Completed
+/// ```
+#[derive(Debug)]
+pub enum PSQState {
+    /// PSQ handshake not yet started.
+    NotStarted,
+
+    /// Initiator has sent PSQ ciphertext and is waiting for confirmation.
+    /// Stores the ciphertext that was sent.
+    InitiatorWaiting { ciphertext: Vec<u8> },
+
+    /// Responder is ready to receive and decapsulate PSQ ciphertext.
+    ResponderWaiting,
+
+    /// PSQ handshake completed successfully.
+    /// The PSK has been derived and registered with the Noise protocol.
+    Completed {
+        /// The derived post-quantum PSK
+        psk: [u8; 32],
+    },
+}
+
+/// A session in the Lewes Protocol, handling connection state with Noise.
+///
+/// Sessions manage connection state, including LP replay protection and Noise cryptography.
+/// Each session has a unique receiving index and sending index for connection identification.
+///
+/// ## PSK Injection Lifecycle
+///
+/// 1. Session created with dummy PSK `[0u8; 32]` in Noise HandshakeState
+/// 2. During handshake, PSQ runs and derives real post-quantum PSK
+/// 3. Real PSK injected via `set_psk()` - `psk_injected` flag set to `true`
+/// 4. Handshake completes, transport mode available
+/// 5. Transport operations (`encrypt_data`/`decrypt_data`) check `psk_injected` flag for safety
+#[derive(Debug)]
+pub struct LpSession {
+    id: u32,
+
+    /// Flag indicating if this session acts as the Noise protocol initiator.
+    is_initiator: bool,
+
+    /// Noise protocol state machine
+    noise_state: Mutex<NoiseProtocol>,
+
+    /// KKT (KEM Key Transfer) exchange state
+    kkt_state: Mutex<KKTState>,
+
+    /// PSQ (Post-Quantum Secure PSK) handshake state
+    psq_state: Mutex<PSQState>,
+
+    /// PSK handle from responder (ctxt_B) for future re-registration
+    psk_handle: Mutex<Option<Vec<u8>>>,
+
+    /// Counter for outgoing packets
+    sending_counter: AtomicU64,
+
+    /// Validator for incoming packet counters to prevent replay attacks
+    receiving_counter: Mutex<ReceivingKeyCounterValidator>,
```

Comment:
I haven't looked exhaustively through all possible codepath, but all mutating data access to the fields behind mutexes are happening via `LpStateMachine::process_input`. and that one does use `&mut self`, thus do we even need the overhead of mutexes at all? they don't seem to be providing any value